### PR TITLE
fixed catalyst to correctly extract number of shots for legacyDeviceFacade 

### PR DESCRIPTION
**Pull request by mehrdad2m**
**Context:**
This PR is to fix an issue that will be caused after merging the following PR:
https://github.com/PennyLaneAI/pennylane/pull/6046

The problem is that LegacyDeviceFacade will cause the Legacy Devices to use the Shots class instead of an integer which causes a crash in runtime.

**Description of the Change:**
This PR ensures that the any device that is an instance of the new device API uses the shots class to extract the total number of shots.

**Related GitHub Issues:**
[sc-70792]

### CODE REVIEW

File: frontend/catalyst/device/qjit_device.py
Comment (dime10) [inline frontend/catalyst/device/qjit_device.py:204]: What scenario does this cover? Is it problematic to reassign the value?

**Review by dime10 — COMMENTED** [review]
Nice work debugging this 💯 

Does that we mean we can give the go ahead on the PennyLane PR and we'll fetch that version once merged in this PR?

File: frontend/catalyst/device/qjit_device.py
Comment (mehrdad2m) [reply frontend/catalyst/device/qjit_device.py:204]: This is for a case where the shots was included in the capabilities.options and this would reassign the shots that was already set by get_device_shots to the unprocessed shots that causes the issue.

**Comment by mehrdad2m** [comment]
> Nice work debugging this 💯
> 
> Does that we mean we can give the go ahead on the PennyLane PR and we'll fetch that version once merged in this PR?

Thanks @dime10! I have just found a bug in that PR which I will report, but I don't know how important that is. Otherwise, it is good to go. 

for your reference this is the bug: 
```
from copy import copy
dev = qml.device("qrack.simulator", wires=1, shots=20)
new_dev = copy(dev)
```

The copy function causes a RecursioError when using a LegacyDevice

```
Traceback (most recent call last):
  File "/Users/mehrdad.malek/tmp/test-new-device.py", line 27, in <module>
    new_dev = copy(dev)
              ^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/copy.py", line 97, in copy
    return _reconstruct(x, None, *rv)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/copy.py", line 260, in _reconstruct
    if hasattr(y, '__setstate__'):
       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/mehrdad.malek/pennylane/pennylane/devices/legacy_facade.py", line 191, in __getattr__
    return getattr(self._device, name)
                   ^^^^^^^^^^^^
  File "/Users/mehrdad.malek/pennylane/pennylane/devices/legacy_facade.py", line 191, in __getattr__
    return getattr(self._device, name)
                   ^^^^^^^^^^^^
  File "/Users/mehrdad.malek/pennylane/pennylane/devices/legacy_facade.py", line 191, in __getattr__
    return getattr(self._device, name)
                   ^^^^^^^^^^^^
  [Previous line repeated 994 more times]
RecursionError: maximum recursion depth exceeded
```

The implication of this in catalyst is a crash when using dynamic one-shot with a legacyDevice

**Comment by dime10** [comment]
> > Nice work debugging this 💯
> > Does that we mean we can give the go ahead on the PennyLane PR and we'll fetch that version once merged in this PR?
> 
> Thanks @dime10! I have just found a bug in that PR which I will report, but I don't know how important that is. Otherwise, it is good to go.
> 
> for your reference this is the bug:
> 
> ```
> from copy import copy
> dev = qml.device("qrack.simulator", wires=1, shots=20)
> new_dev = copy(dev)
> ```
> 
> The copy function causes a RecursioError when using a LegacyDevice


Good find, sounds important to me but maybe the core team will weigh in on the PennyLane repo :)

**Comment by mehrdad2m** [comment]
> > > Nice work debugging this 💯
> > > Does that we mean we can give the go ahead on the PennyLane PR and we'll fetch that version once merged in this PR?
> > 
> > 
> > Thanks @dime10! I have just found a bug in that PR which I will report, but I don't know how important that is. Otherwise, it is good to go.
> > for your reference this is the bug:
> > ```
> > from copy import copy
> > dev = qml.device("qrack.simulator", wires=1, shots=20)
> > new_dev = copy(dev)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The copy function causes a RecursioError when using a LegacyDevice
> 
> Good find, sounds important to me but maybe the core team will weigh in on the PennyLane repo :)

An update on this: 
Ahmed fixed this issue very quickly and it got resolved.

File: frontend/catalyst/qfunc.py
Comment (Shiro-Raven) [inline frontend/catalyst/qfunc.py:262]: This check is redundant. Since the facade will ALWAYS exist around legacy devices, we don't want any special handling for Legacy devices in the wild, as this should be considered a bug.
This branch should be deleted.

**Review by Shiro-Raven — CHANGES_REQUESTED** [review]
LG from core side. I am only requesting the removal of the redundant `elif` in the changes.

File: frontend/catalyst/qfunc.py
Comment (mehrdad2m) [reply frontend/catalyst/qfunc.py]: Thanks @Shiro-Raven for the comment! I was trying to be conservative so it works in a wild case where the target_device is extracted from the facade and used separately, but seems like we don't want that to happen. removed the branch

Comment (Shiro-Raven) [reply frontend/catalyst/qfunc.py]: We tried our best to cover all potential insertions of a device into a circuit execution such that it always gets caught and wrapped with the facade. If a scenario shows up in the future where a user was able to somehow circumvent all the checks we added, I think it would be better to handle it in PL instead of it being silently handled in other libraries.

Comment (mehrdad2m) [reply frontend/catalyst/qfunc.py]: Actually removing that check causes test failures for kokkos. Since kokkos CI job is using version 0.37.0 of lightning, I think maybe it is best to bring the check back and have TODO to remove it after the next release. What do you think @dime10?

Comment (dime10) [reply frontend/catalyst/qfunc.py]: Shouldn't we be using the dev version of Kokkos? Where does this happen?

Comment (mehrdad2m) [reply frontend/catalyst/qfunc.py]: you are right! sorry I looked at the wrong place. It actually uses the dev version. so we are good.

**Review by Shiro-Raven — APPROVED** [review]
LGTM! Thanks!

**Review by erick-xanadu — APPROVED** [review]

**Comment by erick-xanadu** [comment]
Just wait until the tests finish running :)

**Comment by mehrdad2m** [comment]
> Just wait until the tests finish running :)

Actually I can't merge until Ahmed merges his PR first. Otherwise kokkos CI will fail.

**Comment by erick-xanadu** [comment]
> Actually I can't merge until Ahmed merges his PR first. Otherwise kokkos CI will fail.

@Shiro-Raven

**Comment by Shiro-Raven** [comment]
> > Just wait until the tests finish running :)
> 
> Actually I can't merge until Ahmed merges his PR first. Otherwise kokkos CI will fail.

Yes I am aware. I will handle the merges myself. Dw about it :)

**Pull request merged**
